docs/src/pages/demos/lists/SimpleList.tsx

```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { createStyles, Theme, withStyles, WithStyles } from '@material-ui/core/styles';
+import List from '@material-ui/core/List';
+import ListItem, { ListItemProps, ListItemTypeMap } from '@material-ui/core/ListItem';
+import ListItemIcon from '@material-ui/core/ListItemIcon';
+import ListItemText from '@material-ui/core/ListItemText';
+import Divider from '@material-ui/core/Divider';
+import InboxIcon from '@material-ui/icons/Inbox';
+import DraftsIcon from '@material-ui/icons/Drafts';
+
+const styles = (theme: Theme) =>
+  createStyles({
+    root: {
+      width: '100%',
+      maxWidth: 360,
+      backgroundColor: theme.palette.background.paper,
+    },
+  });
+
+interface ListItemLinkProps {
```

Comment:
@pelotom Most of the typescript issues use some sort of props spread. I can't just simple extend `ListItemProps` here. How would I define the widest possible type that can be spread into `<ListItem button component="a" />`?

Reply:
I think something like this would work:

```ts
type ListItemLinkProps = OverrideProps<ListItemTypeMap<{ button?: true;  }, 'button'>, 'a'>;
```

(Note that it shouldn't be necessary to add the `href` prop because that comes by virtue of the fact that you're specifying the override component as `a`.)

This is pretty ugly for users though 🙁 A more user-friendly type operator would perhaps be

```ts
export type ListItemTypeMap2<D extends React.ReactType, P = {}> = ListItemTypeMap<P & D extends 'button'
  ? { button?: true }
  : { button?: false }, D>;
```

Then we can redefine `ListItem` as

```ts
declare const ListItem:
  & OverridableComponent<ListItemTypeMap2<'li'>>
  & OverridableComponent<ListItemTypeMap2<'button'>>;
```

and your `ListItemLinkProps` becomes

```ts
type ListItemLinkProps = OverrideProps<ListItemTypeMap2<'button'>, 'a'>;
```

Still pretty complex, but slightly better?

Reply:
There's just so much ceremony required for this. I'd rather not expose all of these helper types (OverrideProps, ListItemTypeMap etc).

`ListItemProps` should just™ work standalone. They're currently not including e.g. `component`. A better API would be something like `type ListItemLinkProps = ListItemProps<'a'>`. 

`ListItem` is a really hard problem. We're not even covering every possible combination yet :scream: 

Could we try to leverage tagged unions instead of function overloads? 3.2 added non-unit types as union discriminants.

Reply:
Agreed on all points. I tried unions first and they didn't work (for reasons I can't remember specifically now); overloads seemed to be the only thing that worked. But that's probably worth trying again with the most recent version of TypeScript, because I agree that unions would make everything much simpler if they could work.

Reply:
One of the issues is that mapped types don't distribute over unions so interop with other HOCs is pretty bad. This is especially an issue for other styling solution like `styled-components`. They're already not working right now because we use `React.ComponentType` instead of a single type (inferred props are a union because `children` shenanigans). 

I guess we can try to to improve the interop by filing PRs that use `T extends any ? Pick<T, Keys> : never` instead of standard `Pick`.

Reply:
Yeah, I don't know any other solution if we want to make our props accurate, than to help make other libraries' types richer to match. Due to MUI's component overriding pattern the typical component is almost always only expressible as a union, quantified over by a component type parameter. The types are inherently complex because of MUI's extremely flexible design.